In [17]:
import json
import os

In [18]:
folder_path = "./30day"

taipei_data = []
for filename in os.listdir(folder_path):
    print(filename)
    if filename.endswith(".json"):
        file_path = os.path.join(folder_path, filename)
        with open(file_path, "r") as f:
            file_data = f.read()  # .lstrip().rstrip()
            data = json.loads(file_data)
            taipei_data += data["cwaopendata"]["resources"]["resource"]["data"]["surfaceObs"]["location"][3]["stationObsTimes"]["stationObsTime"]


1109-1209.json
1122-1222.json
1029-1128.json
1003-1102.json


In [19]:
import pandas as pd

rain_ar=[]
all_ar=[]
rain_dic={}
Tcnt = 0
Xcnt = 0
Rcnt = 0
T = 0.3 
X = None
for d in taipei_data:
    all = d["weatherElements"]
    rain = d["weatherElements"]["Precipitation"]
    datetime= d["DateTime"]
    if rain == "X":
        Xcnt += 1
        rain = X
    elif rain == "T":
        Tcnt += 1
        rain = T
    else:
        rain = float(rain)
        if rain < 0.0:
            raise ValueError("rain is negative")
        if rain > 0.0:
            Rcnt+=1
        if rain in rain_dic:
            rain_dic[rain]+=1
        else:
            rain_dic[rain]=1
    if "24:00:00" in datetime:
        datetime = datetime.split('T')[0]
        datetime = pd.to_datetime(datetime)+pd.Timedelta(days=1)
    else:
        datetime = pd.to_datetime(datetime).tz_localize(None)
    rain_ar.append({"datetime":datetime,"rain":rain})
    all= {"datetime": datetime,**all}
    all_ar.append(all)
print(f"Total: {len(taipei_data)} (dup)")
print(Tcnt)
print(Rcnt)
print(Xcnt)
print(rain_dic)

Total: 2976 (dup)
438
172
0
{0.0: 2366, 0.5: 94, 2.0: 7, 3.0: 12, 1.0: 19, 2.5: 9, 1.5: 12, 7.5: 3, 9.0: 2, 5.5: 3, 4.0: 4, 7.0: 2, 3.5: 2, 8.5: 1, 5.0: 1, 6.5: 1}


In [20]:
for e in rain_ar:
    print(e)

{'datetime': Timestamp('2023-11-09 01:00:00'), 'rain': 0.0}
{'datetime': Timestamp('2023-11-09 02:00:00'), 'rain': 0.0}
{'datetime': Timestamp('2023-11-09 03:00:00'), 'rain': 0.0}
{'datetime': Timestamp('2023-11-09 04:00:00'), 'rain': 0.0}
{'datetime': Timestamp('2023-11-09 05:00:00'), 'rain': 0.0}
{'datetime': Timestamp('2023-11-09 06:00:00'), 'rain': 0.0}
{'datetime': Timestamp('2023-11-09 07:00:00'), 'rain': 0.0}
{'datetime': Timestamp('2023-11-09 08:00:00'), 'rain': 0.0}
{'datetime': Timestamp('2023-11-09 09:00:00'), 'rain': 0.0}
{'datetime': Timestamp('2023-11-09 10:00:00'), 'rain': 0.0}
{'datetime': Timestamp('2023-11-09 11:00:00'), 'rain': 0.0}
{'datetime': Timestamp('2023-11-09 12:00:00'), 'rain': 0.0}
{'datetime': Timestamp('2023-11-09 13:00:00'), 'rain': 0.0}
{'datetime': Timestamp('2023-11-09 14:00:00'), 'rain': 0.0}
{'datetime': Timestamp('2023-11-09 15:00:00'), 'rain': 0.0}
{'datetime': Timestamp('2023-11-09 16:00:00'), 'rain': 0.0}
{'datetime': Timestamp('2023-11-09 17:00

In [21]:
for e in all_ar:
    print(e)

{'datetime': Timestamp('2023-11-09 01:00:00'), 'AirPressure': '1012.7', 'AirTemperature': '24.9', 'RelativeHumidity': '73', 'WindSpeed': '1.8', 'WindDirection': '東,E', 'Precipitation': '0.0', 'SunshineDuration': None}
{'datetime': Timestamp('2023-11-09 02:00:00'), 'AirPressure': '1012.1', 'AirTemperature': '24.7', 'RelativeHumidity': '74', 'WindSpeed': '1.1', 'WindDirection': '東,E', 'Precipitation': '0.0', 'SunshineDuration': None}
{'datetime': Timestamp('2023-11-09 03:00:00'), 'AirPressure': '1011.9', 'AirTemperature': '24.3', 'RelativeHumidity': '75', 'WindSpeed': '1.0', 'WindDirection': '東南東,ESE', 'Precipitation': '0.0', 'SunshineDuration': None}
{'datetime': Timestamp('2023-11-09 04:00:00'), 'AirPressure': '1012.0', 'AirTemperature': '24.1', 'RelativeHumidity': '77', 'WindSpeed': '0.7', 'WindDirection': '南南西,SSW', 'Precipitation': '0.0', 'SunshineDuration': None}
{'datetime': Timestamp('2023-11-09 05:00:00'), 'AirPressure': '1011.9', 'AirTemperature': '23.5', 'RelativeHumidity': '8

To dataframe 

In [22]:
rain_df = pd.DataFrame(rain_ar)
rain_df.drop_duplicates(subset="datetime",inplace=True)
print(rain_df)


                datetime  rain
0    2023-11-09 01:00:00   0.0
1    2023-11-09 02:00:00   0.0
2    2023-11-09 03:00:00   0.0
3    2023-11-09 04:00:00   0.0
4    2023-11-09 05:00:00   0.0
...                  ...   ...
2851 2023-10-28 20:00:00   0.0
2852 2023-10-28 21:00:00   0.0
2853 2023-10-28 22:00:00   0.0
2854 2023-10-28 23:00:00   0.0
2855 2023-10-29 00:00:00   0.0

[1944 rows x 2 columns]


                       datetime  rain
0     2023-10-03T01:00:00+08:00   0.0
1     2023-10-03T02:00:00+08:00   0.0
2     2023-10-03T03:00:00+08:00   0.0
3     2023-10-03T04:00:00+08:00   0.0
4     2023-10-03T05:00:00+08:00   0.0
...                         ...   ...
2227  2023-12-09T20:00:00+08:00   0.0
2228  2023-12-09T21:00:00+08:00   0.0
2229  2023-12-09T22:00:00+08:00   0.0
2230  2023-12-09T23:00:00+08:00   0.0
2231  2023-12-09T24:00:00+08:00   0.0

[1632 rows x 2 columns]

>>> 1632/(31+30+7)
24.0


In [23]:
all_df = pd.DataFrame(all_ar)
all_df.drop_duplicates(subset="datetime",inplace=True)
print(all_df)

                datetime AirPressure AirTemperature RelativeHumidity  \
0    2023-11-09 01:00:00      1012.7           24.9               73   
1    2023-11-09 02:00:00      1012.1           24.7               74   
2    2023-11-09 03:00:00      1011.9           24.3               75   
3    2023-11-09 04:00:00      1012.0           24.1               77   
4    2023-11-09 05:00:00      1011.9           23.5               80   
...                  ...         ...            ...              ...   
2851 2023-10-28 20:00:00      1013.6           22.9               69   
2852 2023-10-28 21:00:00      1013.7           23.2               66   
2853 2023-10-28 22:00:00      1013.8           22.8               75   
2854 2023-10-28 23:00:00      1013.6           22.6               78   
2855 2023-10-29 00:00:00      1013.0           22.6               79   

     WindSpeed WindDirection Precipitation SunshineDuration  
0          1.8           東,E           0.0             None  
1          

                       datetime AirPressure AirTemperature RelativeHumidity  \
0     2023-10-03T01:00:00+08:00      1008.2           24.9               80   
1     2023-10-03T02:00:00+08:00      1007.5           24.5               82   
2     2023-10-03T03:00:00+08:00      1006.9           24.2               83   
3     2023-10-03T04:00:00+08:00      1006.8           24.1               82   
4     2023-10-03T05:00:00+08:00      1006.9           24.1               77   
...                         ...         ...            ...              ...   
2227  2023-12-09T20:00:00+08:00      1012.1           22.1               85   
2228  2023-12-09T21:00:00+08:00      1011.9           21.7               88   
2229  2023-12-09T22:00:00+08:00      1011.6           21.5               89   
2230  2023-12-09T23:00:00+08:00      1011.2           21.0               91   
2231  2023-12-09T24:00:00+08:00      1010.9           20.6               92   

     WindSpeed WindDirection Precipitation SunshineDuration  
0          1.2         東南,SE           0.0             None  
1          1.5          東,SE           0.0             None  
2          1.4       南南東,SSE           0.0             None  
3          1.2          東,SE           0.0             None  
4          1.8          東,SE           0.0             None  
...        ...           ...           ...              ...  
2227       2.2           東,E           0.0             None  
2228       1.8           東,E           0.0             None  
2229       1.4       東南東,ESE           0.0             None  
2230       0.9          東,SE           0.0             None  
2231       0.8         東南,SE           0.0             None  

[1632 rows x 8 columns]

In [24]:
rain_df.sort_values('datetime',inplace=True)
all_df.sort_values('datetime',inplace=True)
rain_df.to_pickle("./1003-1222_rain.pkl")
all_df.to_pickle("./1003-1222_all.pkl")